src/distdl/nn/batch_norm.py

```diff
@@ -0,0 +1,227 @@
+        The value used for the running_mean and running_var computation.
+        Can be set to None for cumulative moving average (i.e. simple average).
+        Default is 0.1.
+    affine : optional
+        a boolean value that when set to True, this module has learnable affine parameters.
+        Default is True.
+    track_running_stats : optional
+        a boolean value that when set to True, this module tracks the running mean and variance,
+        and when set to False, this module does not track such statistics and uses batch statistics
+        instead in both training and eval modes if the running mean and variance are None.
+        Default is True.
+    """
+
+    def __init__(self, P_x,
+                 num_features, eps=1e-05, momentum=0.1, affine=True,
+                 track_running_stats=True):
+        super(DistributedBatchNorm, self).__init__()
+        self.num_dimensions = len(P_x.shape)
+        assert self.num_dimensions >= 2
+        self.num_features = num_features
+        self.eps = eps
+        self.momentum = momentum
+        self.affine = affine
+        self.track_running_stats = track_running_stats
+        self.inputs_seen = 0
+
+        # Determine the size of the local trainable parameters (this is a bit of a hack)
+        possible_input_shape = P_x.shape.tolist()
+        possible_input_shape[1] = num_features
+        start_index = compute_start_index(P_x.shape, P_x.index, possible_input_shape)
+        stop_index = compute_stop_index(P_x.shape, P_x.index, possible_input_shape)
+        self.local_num_features = stop_index[1] - start_index[1]
+
+        internal_data_shape = [1] * self.num_dimensions
+        internal_data_shape[1] = self.local_num_features
+        internal_partition_shape = [1] * self.num_dimensions
+        internal_partition_shape[1] = P_x.shape[1]
+
+        # Decide which workers will be used to store sum and affine parameters
+        index = [0] * self.num_dimensions
+        index[1] = slice(0, P_x.shape[1])
+        index = tuple(index)
+        storage_workers = worker_layout(P_x.shape)[index].tolist()
+
+        self.P_x = P_x
+        self.P_sum_base = P_x.create_partition_inclusive(storage_workers)
+        self.P_sum = self.P_sum_base.create_cartesian_topology_partition(internal_partition_shape)
+
+        if self.track_running_stats:
+            self.running_mean = torch.zeros(internal_data_shape)
+            self.running_var = torch.ones(internal_data_shape)
+        else:
+            self.running_mean = None
+            self.running_var = None
+
+        self.sr = SumReduce(P_x, self.P_sum)
+        self.bc = Broadcast(self.P_sum, P_x)
+
+        if self.affine:
+            if self.P_sum.active:
+                self.gamma = torch.nn.Parameter(torch.ones(internal_data_shape))
+                self.beta = torch.nn.Parameter(torch.zeros(internal_data_shape))
+            else:
+                # Note: This is a bit of a hack that ensures that all ranks
```

Comment:
Actually, I think there *might* be an outstanding bug with how DistDL works with optimizers at the moment. The following code fails:

```
P_world = MPIPartition(MPI.COMM_WORLD)
P_base = P_world.create_partition_inclusive(np.arange(4))
P_in_out_base = P_world.create_partition_inclusive([0])
P_x = P_base.create_cartesian_topology_partition([1, 2, 2])
P_in_out = P_in_out_base.create_cartesian_topology_partition([1, 1, 1])

net = torch.nn.Sequential(DistributedTranspose(P_in_out, P_x),
                          DistributedFeatureConv2d(P_x,
                                                   in_channels=1,
                                                   out_channels=1,
                                                   kernel_size=[3, 3]),
                          DistributedTranspose(P_x, P_in_out))
dist_parameters = list(net.parameters())
dist_optim = torch.optim.Adam(dist_parameters, lr=1e-1)
```

This code throws the following error: `ValueError: optimizer got an empty parameter list`. I feel like this is reasonable user input. However, this error will not occur for the affine-BN layer, due to the trivial `Parameter`s. Or perhaps should this code be re-written such that the optimizer is only created on ranks with learn-able parameters? I thought the optimizer must be created on all ranks for each rank to participate in the adjoint phase (`dist_optim.step()`)

In the Lenet5 example, I think this problem was avoided with the final Linear layer, which stores weights in all four ranks.

Reply:
I agree entirely.  However, I'm not convinced that this is the correct solution.  I think there is a good argument in its favor, but I also want to consider some other potential approaches before we commit to one.  It is unfortunate that Torch things that there is no reason to compute any gradients if there are no Parameters...

Reply:
Leaving some notes on this for bookkeeping. The plan is to remove the trivial parameters and mimic the other tests--instead of the optimizer and sequential net, we'll use the `backward()` function and the `grad` property of the layer.